src/network/core/tcp_connect.cpp

```diff
@@ -467,24 +462,14 @@ void TCPServerConnecter::SetFailure()
  */
 /* static */ void TCPConnecter::CheckCallbacks()
 {
-	for (auto iter = _tcp_connecters.begin(); iter < _tcp_connecters.end(); /* nothing */) {
-		TCPConnecter *cur = *iter;
-
-		if (cur->CheckActivity()) {
-			iter = _tcp_connecters.erase(iter);
-			delete cur;
-		} else {
-			iter++;
-		}
-	}
+	TCPConnecter::connecters.erase(
+		std::remove_if(TCPConnecter::connecters.begin(), TCPConnecter::connecters.end(),
+				   [](auto &connecter) { return connecter->CheckActivity(); }),
+		TCPConnecter::connecters.end());
 }
 
 /** Kill all connection attempts. */
 /* static */ void TCPConnecter::KillAll()
 {
-	for (auto iter = _tcp_connecters.begin(); iter < _tcp_connecters.end(); /* nothing */) {
-		TCPConnecter *cur = *iter;
-		iter = _tcp_connecters.erase(iter);
-		delete cur;
-	}
+	TCPConnecter::connecters.clear();
```

Comment:
The stupid issue we have now, and this is more what I mentioned earlier, that how we handle connections is gotten so weird in our code, it is hard to untangle. But, the stupid issue: in the old code, the dtor was called when KillAll was called, which causes the connection to be closed. Which could leave `this->game_connecter` in an invalid state.

But now, it will remain open, as there is still a shared-ptr active. An improvement for sure, as there are no dangling things anymore, but it also means `KillAll` isn't actually what it sounds like.

So I wonder if we should just rename this function, or actually address the underlying issue, which requires a bit more rework.

Ugh, every time I look at this part of the code ... can we just CTRL+A DELETE it, and do it again? :D

Reply:
Well, `KillAll` definitely kills all attempts. It just doesn't kill those that were doing the attempts.

I'm not going to worry about the `this->game_connecter`s being in a weird state in this PR, as they cannot be. `KillAll` is only called after the network coordinator client is told to close all connections. That calls `Kill` of the `game_connecter` and closes the turn and stun handler, which were all the classes that had a reference to the `TCPConnecter`.

I'd propose small steps, and leaving the not so desirable situation as-is for this PR. Especially since it's currently not broken due to the way everything is called.